File: xinyu-server/src/main/java/com/xinyu/message/entity/Message.java
package com.xinyu.message.entity;

import com.baomidou.mybatisplus.annotation.FieldFill;
import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableLogic;
import com.baomidou.mybatisplus.annotation.TableName;
import com.xinyu.message.enums.MessageRole;
import com.xinyu.message.enums.MessageStatus;
import lombok.Data;

import java.time.LocalDateTime;

/**
 * 消息实体（对应 scripts/schema.sql 的 message 表, 核心表）
 *
 * <p>枚举列（message_type/status）依赖 MyBatis 默认 EnumTypeHandler
 * 按枚举名读写 VARCHAR, 与数据库字典值严格一致。
 */
@Data
@TableName("message")
public class Message {

    /** 雪花ID（趋势递增, 兼作游标分页的排序键） */
    @TableId(type = IdType.ASSIGN_ID)
    private Long id;

    /** 所属会话 */
    private Long conversationId;

    /** 冗余: 归属用户（统计/权限校验免 join） */
    private Long userId;

    /** 消息树: 重新生成的候选挂同一 USER 消息下（M1-3 暂不使用） */
    private Long parentMessageId;

    /** 会话内自增业务序号（MAX+1, 单用户写无并发问题）, 上下文排序依据 */
    private Integer sequenceNo;

    /** 前端 UUID, 幂等防重复提交, uk(user_id, client_message_id), 可空 */
    private String clientMessageId;

    /** USER / ASSISTANT / SYSTEM */
    private MessageRole messageType;

    /** Markdown 原文 */
    private String content;

    /** GENERATING / COMPLETED / FAILED / STOPPED */
    private MessageStatus status;

    /** 输入 token（仅 ASSISTANT） */
    private Integer promptTokens;

    /** 输出 token（仅 ASSISTANT） */
    private Integer completionTokens;

    /** 实际使用模型（溯源） */
    private String modelCode;

    /** NONE / LIKE / DISLIKE（M1-3 走数据库默认 NONE） */
    private String feedback;

    /** 重新生成次数 */
    private Integer regenerateCount;

    /** 创建时间（插入时自动填充） */
    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /** 更新时间（插入/更新时自动填充） */
    @TableField(fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /** 逻辑删除：0-正常 1-已删除 */
    @TableLogic
    private Integer deleted;
}

File: xinyu-backend/app/models/message.py
"""消息实体 — message 表 (核心表)"""

from datetime import datetime

from sqlalchemy import BigInteger, DateTime, Index, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.security import now_local
from app.models.base import Base, id_primary_key


class Message(Base):
    __tablename__ = "message"
    __table_args__ = (
        UniqueConstraint("user_id", "client_message_id", name="uk_client_msg"),
        Index("idx_conv", "conversation_id", "id"),
        Index("idx_user_created", "user_id", "created_at"),
    )

    id: Mapped[int] = id_primary_key()
    conversation_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    user_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    parent_message_id: Mapped[int | None] = mapped_column(BigInteger)
    sequence_no: Mapped[int] = mapped_column(Integer, nullable=False)
    client_message_id: Mapped[str | None] = mapped_column(String(64))
    message_type: Mapped[str] = mapped_column(String(20), nullable=False, comment="USER/ASSISTANT/SYSTEM")
    content: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="COMPLETED", server_default="COMPLETED"
    )
    prompt_tokens: Mapped[int | None] = mapped_column(Integer)
    completion_tokens: Mapped[int | None] = mapped_column(Integer)
    model_code: Mapped[str | None] = mapped_column(String(50))
    feedback: Mapped[str] = mapped_column(String(20), nullable=False, default="NONE", server_default="NONE")
    regenerate_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=now_local, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=now_local, onupdate=now_local, server_default=func.now()
    )
    deleted: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
